src/qsort.rs

```diff
@@ -0,0 +1,332 @@
+
+	for start in (0..=last_parent).rev() {
+		heap_sift_down(base, start, nel - 1, width, comp);
+	}
+}
+
+fn heap_sift_down(
+	base: *mut CChar,
+	start: CSizeT,
+	end: CSizeT,
+	width: CSizeT,
+	comp: extern "C" fn(*const CVoid, *const CVoid) -> CInt,
+) {
+	// get the left child of the node at the given index
+	let left_child = |idx| 2 * idx + 1;
+
+	let mut root = start;
+
+	while left_child(root) <= end {
+		let child = left_child(root);
+		let mut swap_idx = root;
+
+		let root_ptr = unsafe { base.add(root * width) };
+		let mut swap_ptr = unsafe { base.add(swap_idx * width) };
+		let first_child_ptr = unsafe { base.add(child * width) };
+		let second_child_ptr = unsafe { base.add((child + 1) * width) };
+
+		if comp(swap_ptr as *const CVoid, first_child_ptr as *const CVoid) < 0 {
+			swap_idx = child;
+			swap_ptr = first_child_ptr;
+		}
+		if child < end && comp(swap_ptr as *const CVoid, second_child_ptr as *const CVoid) < 0 {
+			swap_idx = child + 1;
+			swap_ptr = second_child_ptr;
+		}
+
+		if swap_idx == root {
+			break;
+		} else {
+			swap(root_ptr, swap_ptr, width);
+			root = swap_idx;
+		}
+	}
+}
+
+#[inline]
+fn partition(
+	base: *mut CChar,
+	nel: CSizeT,
+	width: CSizeT,
+	comp: extern "C" fn(*const CVoid, *const CVoid) -> CInt,
+) -> (CSizeT, CSizeT) {
+	// calculate the median of the first, middle, and last elements and use it as the pivot
+	// to do fewer comparisons, also swap the elements into their correct positions
+	let mut pivot = median_of_three(base, nel, width, comp);
+
+	let mut i = 1;
+	let mut j = 1;
+	let mut n = nel - 2;
+
+	// use this to deal with the Dutch national flag problem
+	while j <= n {
+		let i_ptr = unsafe { base.add(i * width) };
+		let j_ptr = unsafe { base.add(j * width) };
+		let n_ptr = unsafe { base.add(n * width) };
+		let pivot_ptr = unsafe { base.add(pivot * width) };
+
+		let comparison = comp(j_ptr as *const CVoid, pivot_ptr as *const CVoid);
+		match comparison.cmp(&0) {
+			core::cmp::Ordering::Less => {
+				swap(i_ptr, j_ptr, width);
+				if i == pivot {
+					pivot = j;
+				}
+				i += 1;
+				j += 1;
+			}
+			core::cmp::Ordering::Greater => {
+				swap(j_ptr, n_ptr, width);
+				if n == pivot {
+					pivot = j;
+				}
+				n -= 1;
+			}
+			core::cmp::Ordering::Equal => {
+				j += 1;
+			}
+		}
+	}
+
+	(i, n)
+}
+
+fn median_of_three(
+	base: *mut CChar,
+	nel: CSizeT,
+	width: CSizeT,
+	comp: extern "C" fn(*const CVoid, *const CVoid) -> CInt,
+) -> CSizeT {
+	let pivot = nel / 2;
+
+	let mid = unsafe { base.add(pivot * width) };
+	let last = unsafe { base.add((nel - 1) * width) };
+	if comp(mid as *const CVoid, base as *const CVoid) < 0 {
+		swap(mid, base, width);
+	}
+	if comp(last as *const CVoid, mid as *const CVoid) < 0 {
+		swap(mid, last, width);
+		if comp(mid as *const CVoid, base as *const CVoid) < 0 {
+			swap(mid, base, width);
+		}
+	}
+
+	pivot
+}
+
+#[inline]
+fn swap(mut ptr1: *mut CChar, mut ptr2: *mut CChar, mut width: CSizeT) {
```

Comment:
Isn't this just <https://doc.rust-lang.org/stable/std/ptr/fn.swap.html#examples>, with pointers to arrays of [u8; width] as the args?

Reply:
No, wait. That function needs the size as a constant, and we only have it as a variable.

Reply:
Yes, I would have used `core::slice::sort_unstable_by()` if so.